src/vos/vos_pool_scrub.c

```diff
@@ -0,0 +1,735 @@
+{
+	if (type == VOS_ITER_RECX &&
+	    !recx_eq(ctx->sc_iod.iod_recxs, &entry->ie_recx))
+		return false;
+	return epoch_eq(ctx->sc_epoch, entry->ie_epoch) &&
+	       epoch_eq(ctx->sc_minor_epoch, entry->ie_minor_epc);
+}
+
+static void
+sc_obj_value_reset(struct  scrub_ctx *ctx)
+{
+	ctx->sc_epoch = 0;
+	ctx->sc_minor_epoch = 0;
+}
+
+/** vos_iter_cb_t */
+static int
+obj_iter_scrub_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		  vos_iter_type_t type, vos_iter_param_t *param,
+		  void *cb_arg, unsigned int *acts)
+{
+	struct scrub_ctx	*ctx = cb_arg;
+	int			 rc;
+
+	if (sc_cont_is_stopping(ctx)) {
+		C_TRACE("Container is stopping.");
+		return 1;
+	}
+
+	if (ctx->sc_pool->sp_scrub_sched == DAOS_SCRUB_SCHED_OFF) {
+		C_TRACE("scrubbing is off now");
+		return 1;
+	}
+
+	switch (type) {
+	case VOS_ITER_OBJ:
+		if (oids_are_same(ctx->sc_cur_oid, entry->ie_oid)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_cur_oid, 0, sizeof(ctx->sc_cur_oid));
+		} else {
+			ctx->sc_cur_oid = entry->ie_oid;
+		}
+		break;
+	case VOS_ITER_DKEY:
+		if (keys_are_same(ctx->sc_dkey, entry->ie_key)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_dkey, 0, sizeof(ctx->sc_dkey));
+		} else {
+			ctx->sc_dkey = param->ip_dkey;
+		}
+		break;
+	case VOS_ITER_AKEY:
+		if (keys_are_same(ctx->sc_iod.iod_name, entry->ie_key)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_iod, 0, sizeof(ctx->sc_iod));
+		} else {
+			ctx->sc_iod.iod_name = param->ip_akey;
+		}
+		break;
+	case VOS_ITER_SINGLE:
+	case VOS_ITER_RECX: {
+		if (sc_value_has_been_seen(ctx, entry, type)) {
+			sc_obj_value_reset(ctx);
+		} else {
+			C_TRACE("Scrubbing akey: "DF_KEY", type: %s, rec size: "
+					DF_U64", extent: "DF_RECX"\n",
+				DP_KEY(&param->ip_akey),
+				(type == VOS_ITER_RECX) ? "ARRAY" : "SV",
+				entry->ie_rsize,
+				DP_RECX(entry->ie_orig_recx)
+			);
+
+			sc_obj_val_setup(ctx, entry, type, param, ih);
+
+			rc = sc_verify_obj_value(ctx, &entry->ie_biov, ih);
+			if (ctx->sc_did_yield) {
+				*acts |= VOS_ITER_CB_YIELD;
+				ctx->sc_did_yield = false;
+			}
+
+			if (rc != 0) {
+				D_ERROR("Error Verifying:"DF_RC"\n", DP_RC(rc));
+				return rc;
+			}
+			ds_scrub_sched_control(ctx);
+		}
+		break;
+	}
+
+	case VOS_ITER_DTX:
+	case VOS_ITER_COUUID:
+	case VOS_ITER_NONE:
+		D_ASSERTF(0, "Invalid.");
+	}
+
+	return 0;
+}
+
+static int
+sc_scrub_cont(struct scrub_ctx *ctx)
+{
+	vos_iter_param_t		param = {0};
+	struct vos_iter_anchors		anchor = {0};
+	int				rc;
+
+	/* not all containers in the pool will have checksums enabled */
+	if (!daos_csummer_initialized(sc_csummer(ctx)))
+		return 0;
+
+	C_TRACE("Scrubbing container '"DF_UUIDF"'\n",
+		DP_UUID(*sc_cont_uuid(ctx)));
+
+	param.ip_hdl = sc_cont_hdl(ctx);
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+	rc = vos_iterate(&param, VOS_ITER_OBJ, true, &anchor,
+			 obj_iter_scrub_pre_cb,
+			 NULL, ctx, NULL);
+
+	if (rc != DER_SUCCESS) {
+		D_ERROR("Object scrub failed: "DF_RC"\n", DP_RC(rc));
+		return rc;
```

Comment:
Differentiate real error, container stopping or scrubbing off. And should we wait a while then continue scrubbing if any error happened?

Reply:
Fixed error handling. Also added a minute wait at ULT level if ds_scrub_pool returns error.